examples/tv-casting-app/android/App/app/src/main/jni/com/chip/casting/TvCastingApp.java

```diff
@@ -136,30 +146,119 @@ public void discoverVideoPlayerCommissioners(
+      List<VideoPlayer> cachedVideoPlayers,
+      List<DiscoveredNodeData> discoveredPlayers,
+      SuccessCallback<DiscoveredNodeData> discoverySuccessCallback) {
+    Log.d(
+        TAG,
+        "TvCastingApp.reportSleepingVideoPlayerCommissioners called with commissioner count "
+            + (cachedVideoPlayers != null ? cachedVideoPlayers.size() : 0));
+    if (cachedVideoPlayers != null) {
+      for (VideoPlayer player : cachedVideoPlayers) {
+        Log.d(TAG, "Cached Video Player: " + player);
+        // do NOT surface this cached Player if we don't have its MACAddress
+        if (player.getMACAddress() == null) {
+          Log.d(
+              TAG,
+              "TvCastingApp.reportSleepingVideoPlayerCommissioners Skipping Player with hostName"
+                  + player.getHostName()
+                  + " but no MACAddress");
+          continue;
+        }
+
+        // do NOT surface this cached Player if it has not been discoverable recently (in
+        // CHIP_DEVICE_CONFIG_STR_CACHE_LAST_DISCOVERED_HOURS)
+        if (!WasRecentlyDiscoverable(player)) {
+          Log.d(
+              TAG,
+              "TvCastingApp.reportSleepingVideoPlayerCommissioners Skipping Player with hostName"
+                  + player.getHostName()
+                  + " that has not been discovered recently");
+          continue;
+        }
+
+        // do NOT surface this cached Player if it was just discovered right now (in this discovery
+        // call)
+        boolean justDiscovered =
+            discoveredPlayers
+                .stream()
+                .anyMatch(
+                    new Predicate<DiscoveredNodeData>() {
+                      @Override
+                      public boolean test(DiscoveredNodeData discoveredNodeData) {
+                        return player.getHostName().equals(discoveredNodeData.getHostName());
+                      }
+                    });
+        if (justDiscovered) {
+          Log.d(
+              TAG,
+              "TvCastingApp.reportSleepingVideoPlayerCommissioners Skipping Player with hostName"
+                  + player.getHostName()
+                  + " that was just discovered");
+          continue;
+        }
+
+        // DO surface this cached Player (as asleep)
+        Log.d(TAG, "Reporting sleeping player with hostName " + player.getHostName());
+        player.setIsAsleep(true);
+        discoverySuccessCallback.handle(new DiscoveredNodeData(player));
+      }
+    }
+  }
+
+  private native boolean WasRecentlyDiscoverable(VideoPlayer player);
+
   public void stopVideoPlayerDiscovery() {
     synchronized (discoveryLock) {
       Log.d(TAG, "TvCastingApp trying to stop video player discovery");
       if (this.discoveryStarted
           && nsdManager != null
           && multicastLock != null
-          && nsdDiscoveryListener != null) {
+          && nsdDiscoveryListener != null
+          && reportSleepingVideoPlayerCommissionersFuture != null) {
```

Comment:
i don't think you need this null check for this

you probably only need the null check on line 276 when you're actually trying to cancel it

Reply:
Moved it.